include/pybind11/pybind11.h

```diff
@@ -462,6 +462,7 @@ class cpp_function : public function {
             pybind11_fail("Internal error while parsing type signature (2)");
         }
 
+        signature.erase(std::remove(signature.begin(), signature.end(), '\n'), signature.end());
```

Comment:
Could this be more targeted?

If the newlines come in through default values, could this be a better place to strip them out (currently line 427)?

```
signature += rec->args[arg_index].descr;
```

I don't think just stripping out newlines is safe in the general case. That might join two "words" into one, which could be super misleading (you have assume that a human glancing through a pile of docstrings has little to no domain knowledge, and we don't want to confuse them). I think we need something more sophisticated, roughly, replace newlines with spaces, but squash repeats, and strip off leading and trailing spaces.

But what we need foremost: test cases, added under pybind11/tests. That's probably 99+% of the work needed here, but a strict requirement for a core tool like this one.

I recommend starting we one test case, work on the implementation, add another test case, refine the implementation, etc. until the test suite covers all corner cases.

Reply:
Thank you for the quick response. I expected as much, but wasn't quite sure where to even begin I'll start looking into it.

Reply:
You seem to have embraced doing the implementation first, which makes sense as long as it remains relatively simple.

If you push the implementation to the point that all existing tests pass, I could run Google's global testing to see if anything gets upset (I'd guess not, but I've been weirdly surprised before several times; changing docstrings generally gets me worried more than other things).

I.e. the idea is:

* Please let me know when you have all tests working here, but maybe hold off writing the tests.
* I run global testing (takes a lot of computing resources, but not a lot of my time).
* If there are no surprises you add the tests.